akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -548,3 +545,228 @@ final private[stream] class LazySink[T, M](sinkFactory: T ⇒ Future[Sink[T, M]]
+              pull(in)
+            case Supervision.Resume ⇒
+              pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = closeStage()
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        close(resource)
+        doFailStage(ex)
+      }
+
+      private def restartState(): Unit = {
+        close(resource)
+        resource = open()
+      }
+
+      private def closeStage(): Unit =
+        try {
+          close(resource)
+          doCompleteStage()
+        } catch {
+          case NonFatal(ex) ⇒ doFailStage(ex)
+        }
+
+      private def doFailStage(th: Throwable): Unit = {
+        completion.failure(th)
+        failStage(th)
+      }
+
+      private def doCompleteStage(): Unit = {
+        completion.success(Done)
+        completeStage()
+      }
+    }
+
+    (logic, completion.future)
+  }
+
+  override def toString = "FoldResourceSink"
+}
+
+/**
+ * INTERNAL API
+ */
+final class FoldResourceSinkAsync[T, S](
+  open:      () ⇒ Future[S],
+  writeData: (S, T) ⇒ Future[Unit],
+  close:     (S) ⇒ Future[Unit])
+  extends GraphStageWithMaterializedValue[SinkShape[T], Future[Done]] {
+  val in = Inlet[T]("FoldResourceSinkAsync.out")
+  override val shape = SinkShape(in)
+  override def initialAttributes: Attributes = DefaultAttributes.foldResourceSinkAsync
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val completion = Promise[Done]()
+    val logic: GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+      var resource = Promise[S]()
+      var writeFuture: Future[Unit] = Future.successful(())
+
+      implicit val context = ExecutionContexts.sameThreadExecutionContext
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        createStream()
+        setKeepGoing(true)
+      }
+
+      private def createStream(): Unit = {
+        val cb = getAsyncCallback[Try[S]] {
+          case scala.util.Success(res) ⇒
+            resource.success(res)
+            if (!isClosed(in)) {
+              pull(in)
+            }
+          case scala.util.Failure(t) ⇒ doFailStage(t)
+        }
+        try {
+          open().onComplete(cb.invoke)
+        } catch {
+          case NonFatal(ex) ⇒ doFailStage(ex)
+        }
+      }
+
+      private def onResourceReady(f: (S) ⇒ Unit): Unit = {
+        val cb = getAsyncCallback[S](f)
+        resource.future.onSuccess {
+          case r ⇒ cb.invoke(r)
+        }
+      }
+
+      private def onWriteComplete(f: Try[Unit] ⇒ Unit): Unit = {
+        val cb = getAsyncCallback[Try[Unit]](f)
+        writeFuture.onComplete(cb.invoke)
+      }
+
+      val errorHandler: PartialFunction[Throwable, Unit] = {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop    ⇒ closeAndThen(() ⇒ doFailStage(ex))
```

Comment:
I feel handling exceptions by passing around as an argument is difficult because read()/write()/close() and onUpstreamFailure() race. Maybe some other race would exist.
I changed the exception handling to accumulate all exceptions to an ArrayBuffer, then check after every close().